### ISSUE TITLE
Non-standard Softmax behavior

### ISSUE THREAD

**Issue by JamesAllingham** [onnx/onnx#2289]
What is the reason for the non-standard behaviour of the ONNX `Softmax` operator (also `LogSoftmax` and `HardMax`)? More specifically, why do we limit (or coerce) the input to be a 2D vector? Many other deep learning frameworks ([TensorFlow](https://www.tensorflow.org/api_docs/python/tf/nn/softmax), [PyTorch](https://pytorch.org/docs/stable/nn.html#softmax), [MXNet](http://mxnet.incubator.apache.org/versions/1.5.0/api/python/ndarray/ndarray.html?highlight=softmax#mxnet.ndarray.softmax)) allow the input of the softmax to be N-dimensional and use the `axis` argument to specify which dimension of the input should be softmaxed (such that the sum of the elements in that dimension is 1). 

It seems to me like the current behaviour is less flexible than the N-dimensional behaviour, and ends up requiring tricks like first transposing the input before applying softmax (as seen in #1543). I also find it a bit strange that an activation function changes the shape of its input.

I propose that we standardise the ONNX Softmax op to follow the behaviour of the deep learning frameworks mentioned above. If the existing behaviour is needed for some applications, it can easily be recovered by reshaping tensors to be 2D. As far as I know, this is a much less common use-case (it isn't implemented in the most popular frameworks) so this shouldn't affect too many users. 

I am more than happy to make a PR for these changes, but I wanted to find out if I was missing some key insight into why have the current behaviour before making any changes.

**Comment by WilliamTambellini** [comment]
I would also be in favor for the input to be possibly n-dimensional.
Though I would remind that the usual NN backends are not that flexible on the softmax axis choice : 
example of cudnn :
https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnSoftmaxForward
https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnSoftmaxMode_t
or mkldnn: 
https://intel.github.io/mkl-dnn/dev_guide_softmax.html

**Comment by wschin** [comment]
> If ONNX Transpose + ONNX Softaxmax can achieve general Softmax, I'd say the expressiveness of ONNX is enough, and we probably don't need a spec change. Note that the graph stored and the graph gets executed can be different. Runtime engine can fuse ONNX Transpose + ONNX Softmax into its general Softmax for better performance.
> 
> While ONNX Transpose + ONNX Softxmax can achieve general Softmax in most cases, some high-level deep learning frameworks abstract away the batch dimension from the construction of the net (for example Keras, Matlab, Mathematica), in these cases ONNX Transpose + ONNX Softxmax is not general enough, because it requires modifying the batch dimension.
> Additionally, although the graph stored and the graph executed can be different, it makes sense to make have general Softmax in ONNX because it will reduce the work for runtime engine implementers (as they don't have to add optimisations for ONNX Transpose + ONNX Softaxmax). There are many runtime engines around and will likely be many more in future. We can make a single change to ONNX that improves performance, or we can leave it up to each individual runtime engine implementation to improve performance, which requires more work and might never actually happen. I think it makes sense to do the work once, even if it requires some effort due to BC-breaking changes.

My feeling here is kind of counter to my opinions [here](https://github.com/onnx/onnx/pull/2330#issuecomment-532754300), because I feel softmax is usually not a computation bottleneck. We have `Gemm` because it's the heaviest part in almost every neural network model and we want every runtime to optimize Gemm.

**Comment by JamesAllingham** [comment]
> 
> 
> > If ONNX Transpose + ONNX Softaxmax can achieve general Softmax, I'd say the expressiveness of ONNX is enough, and we probably don't need a spec change. Note that the graph stored and the graph gets executed can be different. Runtime engine can fuse ONNX Transpose + ONNX Softmax into its general Softmax for better performance.
> > While ONNX Transpose + ONNX Softxmax can achieve general Softmax in most cases, some high-level deep learning frameworks abstract away the batch dimension from the construction of the net (for example Keras, Matlab, Mathematica), in these cases ONNX Transpose + ONNX Softxmax is not general enough, because it requires modifying the batch dimension.
> > Additionally, although the graph stored and the graph executed can be different, it makes sense to make have general Softmax in ONNX because it will reduce the work for runtime engine implementers (as they don't have to add optimisations for ONNX Transpose + ONNX Softaxmax). There are many runtime engines around and will likely be many more in future. We can make a single change to ONNX that improves performance, or we can leave it up to each individual runtime engine implementation to improve performance, which requires more work and might never actually happen. I think it makes sense to do the work once, even if it requires some effort due to BC-breaking changes.
> 
> My feeling here is kind of counter to my opinions [here](https://github.com/onnx/onnx/pull/2330#issuecomment-532754300), because I feel softmax is usually not a computation bottleneck. We have `Gemm` because it's the heaviest part in almost every neural network model and we want every runtime to optimize Gemm.

You are right that optimizing `Softmax` is not nearly as useful or important as optimizing `Gemm`, so I'll happily concede that point :) 

However, I still think that for DL frameworks which abstract away the batch dimension it is beneficial to have the more general version of `Softmax`. 

Is there a strong reason not to make the change? Here are the pros and cons as I see them. 
Pros:

1. ONNX will have a more general and flexible `Softmax` operator. This will make it easier to represent some nets and will also make it easier for some converter implementors (for example for frameworks which abstract away the batch dim).

2. (minor) It may be more efficient in some cases.

Cons:

1. This is a BC-breaking change, which will require some work from backend implementations to support. (Of course, I would add converters to upgrade and downgrade the `Softmax` op where appropriate.)

2. Some backends won't be able to support the more general `Softmax`? I'm not sure if this is actually the case, or if it is actually a problem (since I imagine we don't want to limit the expressiveness of ONNX because some backends aren't as flexible as others).

As I've mentioned, I am more than happy to make the required changes on the ONNX side, so I hope that the work for the ONNX implementors isn't a primary concern here.

All of this said, I will defer to those of you who have been working on ONNX for a while, as I'm sure you have a better idea than me which changes are worth it and which aren't. I'm just trying to understand this situation here :)

**Comment by WilliamTambellini** [comment]
I am challenging these quotes : 
"You are right that optimizing Softmax is not nearly as useful or important as optimizing Gemm, so I'll happily concede that point :)"
Even if for most image classification NN the number of classes is pretty 'small' (1000 classes for the original ResNet if I remember right), the number of classes for some NLP NN could be very high : see several dozen of thousand of classes in order to cover the vocabulary of most human languages.
Ref: see a 32k vocab size for instance there: 
https://github.com/tensorflow/nmt
In real life, I have simply seen softmax (well logsoftmax, fwd and/or bwd) reaching the top 3 bottleneck finalists.

**Comment by wschin** [comment]
@WilliamTambellini, you're right. In extreme classification, SoftMax could be a bottleneck. Do you have prof number for the nmt link you shared? I also remember that due to the huge number of classes, hierarchical SoftMax (or other advanced tricks) were proposed to reduce the SoftMax's computation aomunt. If those tricks are considered, would SoftMax still be bottleneck?

**Comment by daquexian** [comment]
@wschin I also think that the current softmax behavior should be updated.

1. Not all general softmax can be implemented by transpose + onnx softmax, for example, in some cases, a much more complicated graph (reducemax + sub + exp + reducesum + div) has to be produced (https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset9.py#L645-L654) which is much harder for backend optimizers to optimize.

2. As said by @JamesAllingham , it is a breaking change, but onnx has multiple opsets, breaking changes are not a problem at all.

3. Most important, ONNX Softmax doesn't align with almost every mainstream framework ([tensorflow](https://www.tensorflow.org/api_docs/python/tf/nn/softmax), [pytorch](https://pytorch.org/docs/master/generated/torch.nn.Softmax.html), [mxnet](http://beta.mxnet.io/r/api/mx.nd.softmax.html) and [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html)). It hurts the interoperability and brings no benefit. Why do we stick to the current behavior?

**Comment by scarsty** [comment]
菊花残，满地伤，你的笑容已泛黄

**Comment by linkerzhang** [comment]
> @wschin I also think that the current softmax behavior should be updated.
> 
> 1. Not all general softmax can be implemented by transpose + onnx softmax, for example, in some cases, a much more complicated graph (reducemax + sub + exp + reducesum + div) will be produced (https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset9.py#L645-L654) which is much harder for backend optimizers to optimize.
> 2. As said by @JamesAllingham , it is a breaking change, but onnx has multiple opsets, breaking changes are not a problem at all.
> 3. Most important, the current behavior doesn't align with almost every mainstream framework ([tensorflow](https://www.tensorflow.org/api_docs/python/tf/nn/softmax), [pytorch](https://pytorch.org/docs/master/generated/torch.nn.Softmax.html), [mxnet](http://beta.mxnet.io/r/api/mx.nd.softmax.html) and [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html)). It hurts the interoperability and brings no benefit. Why do we stick to the current behavior?

Instead of changing the spec of existing softmax (viewing ONNX as one IR layer), Is adding a "function" (compose operator) working?

ONNX itself should NOT be just one "IR" (layer/spec), but multiple layers (with "function" concept). One "IR" layer will make ONNX either not enough for more and more scenarios or too flatten to be followed by both hardware and software vendors.

**Comment by linkerzhang** [comment]
Adding more, if adding a function does work, then the question is "whether the function should be added into ONNX or added in partner side?"

This question can be answered by how common/general of the scenarios and discussed by operator sig then.

**Comment by ebarsoum** [comment]
We discussed this thread in the operator sig and we agree on the conclusion. As an FYI for the history of this Softmax, it is based on Caffe2 Softmax (https://caffe2.ai/docs/operators-catalogue.html#softmax).

@daquexian the proposal of the function is to write it based on other primitives and not using current Softmax. Is it complicated or not possible?

**Comment by nihui** [comment]
Caffe2's definition of softmax is inherently wrong. Now that caffe2 is gone, almost no one uses caffe2. Maintaining the definition compatibility of softmax leads to unnecessary difficulties for all other framework conversions, and it is already happening now. We had to convert pytorch softmax to a bunch of glue operators in onnx, and then merge these glue operators in the backend.

In response to this softmax problem, the simplest solution to update the definition of softmax is to abandon caffe2 compatibility and bring the friendliness of other framework conversions. In order to multi-level and deliberately multi-level is wrong, softmax is a very commonly used operator, like convolution, pooling, providing a reasonable definition is the main responsibility of onnx.

Regarding the onnx function, I personally don't like multi-level onnx at all, because for front-end and back-end frameworks, it doesn't need that low-level. Whether it is ncnn, tensorRT or openvino, most of them are optimized based on monolithic operator to accelerate. Onnx is currently missing or has a wrong definition of monolithic operator, such as groupnorm, channelshuffle, softmax, etc., which brings trouble to the conversion between frontends and backends. I once wanted to create a new onnx derivative project to fix these missing and inconvenient operator definitions, but I think it is better to fix it in onnx.

Although I don't like the onnx function, I admit that it's not bad, it brings great scalability and compatibility. Even if softmax is implemented using function, is this monolithic softmax definition still to be updated? In any case, we need to update the definition, then update the definition first.

**Comment by daquexian** [comment]
> @daquexian the proposal of the function is to write it based on other primitives and not using current Softmax. Is it complicated or not possible?

@ebarsoum In this case (not using current Softmax) we can always write softmax function as https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic_opset9.py#L645-L654, so it is ok. 

And I have the same problem with @nihui . Will the existing softmax operator be removed and replaced with the new softmax function?

**Comment by ebarsoum** [comment]
@nihui I am not saying that Caffe2 implementation is right, softmax was in the initial set of operators in ONNX and was based on Caffe2, I am just providing a background on why it was written this way. The goal of Function is to reduce the number of primitive OPs in ONNX and enable runtime to optimize them if they need so.

@daquexian and @nihui the two options that I see:

1. Turn Softmax to a Function with the common signature and behavior and deprecate the old Softmax. This is the recommended approach.
2. Update existing Softmax to the new signature and bump its version number.

**Comment by daquexian** [comment]
> @daquexian and @nihui the two options that I see:
> 
> 1. Turn Softmax to a Function with the common signature and behavior and deprecate the old Softmax. This is the recommended approach.
> 2. Update existing Softmax to the new signature and bump its version number.

I also prefer option 1.
